### ISSUE TITLE
How to add AdEventListener with exo player instance?

### ISSUE THREAD

**Issue by AdityaSinghShekhawat** [androidx/media#1681]
Hey, we are updating Exoplayer to media3 exoplayer 1.3.1. For ads we are using Exoplayer IMA extension. We want to execute player commands as per ads events. As per [official docs](https://developers.google.com/interactive-media-ads/docs/sdks/android/client-side/exoplayer-extension) we can add ad event listeners while building `ImaAdsLoader` which is then used in initializing the exo player. How can we pass exo player instance after buliding the`ImaAdsLoader`? FYI - We can't use `onAdsManagerLoaded` as it never gets triggered ([Ref](https://github.com/google/ExoPlayer/issues/9453#issuecomment-924865612)). Ultimately we are not able to attach AdEventListener after exoplayer is instantiated. 

**MyActivity.java**
```
public class MyActivity extends Activity {
  private ExoPlayer player;
  private ImaAdsLoader adsLoader;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    // ...
    // Create an AdsLoader.
    adsLoader =
        new ImaAdsLoader.Builder(this)
            .setAdEventListener(buildAdEventListener())
            .build();
  }

  public AdEvent.AdEventListener buildAdEventListener() {
    AdEvent.AdEventListener imaAdEventListener =
        event -> {
          AdEvent.AdEventType eventType = event.getType();
          if (eventType == AdEvent.AdEventType.CONTENT_PAUSE_REQUESTED) {
            // Use exoplayer to change playback speed to normal right before an ad is about to cover the content.
           // But exoplayer is null because exoplayer requires adsLoader to get initialized.
          } else if (eventType == AdEvent.AdEventType.CONTENT_RESUME_REQUESTED) {
           // Use exoplayer to change playback speed to current video speed when an ad finishes or collapses.
           // But exoplayer is null because exoplayer requires adsLoader to get initialized.
          }
    return imaAdEventListener;
  }

  private void initializePlayer() {
    // Set up the factory for media sources, passing the ads loader and ad view providers.
    DataSource.Factory dataSourceFactory = new DefaultDataSource.Factory(this);

    MediaSource.Factory mediaSourceFactory =
        new DefaultMediaSourceFactory(dataSourceFactory)
            .setLocalAdInsertionComponents(unusedAdTagUri -> adsLoader, playerView);

    // Create an ExoPlayer and set it as the player for content and ads.
    player = new ExoPlayer.Builder(this).setMediaSourceFactory(mediaSourceFactory).build();
    playerView.setPlayer(player);
    adsLoader.setPlayer(player);
}
```

**Comment by marcbaechinger** [comment]
Hmmm, not sure I understand the problem.

From your code above, and where you have the comments `// Use exoplayer to change playback speed `, why can't you access the `MyActivivty.player` there? I mean, the field probably needs to be `@Nullable` but it should be possible to do something like this (adapting your code):

```
public AdEvent.AdEventListener buildAdEventListener() {
    AdEvent.AdEventListener imaAdEventListener =
        event -> {
          assertNotNull(player);
          AdEvent.AdEventType eventType = event.getType();
          if (eventType == AdEvent.AdEventType.CONTENT_PAUSE_REQUESTED) {
            contentPlaybackParameters = player.getPlaybackParameters()
            player.setPlaybackParameter(new PlaybackParameters(...,...));
          } else if (eventType == AdEvent.AdEventType.CONTENT_RESUME_REQUESTED) {
            player.setPlaybackParameters(contentPlaybackParameters);
          }
    return imaAdEventListener;
  }
```

**Comment by AdityaSinghShekhawat** [comment]
Hey @marcbaechinger,

Thanks for reply and apologies for late response. I may not be able to explain my problem clearly. We need to attach multiple `adEventListeners`. Earlier we use to attach `adEventListeners` using ads manager `addAdEventListener`. Now, as ads manager is not available and only place to attach adEventListener is when building ads loader and using `setAdEventListener`. Only downside of this way is that `setAdEventListener` accepts single adEventListener and we have multiple adEventListeners. As a solution I have created CompositeAdEventListener as below and passed it to `setAdEventListener`.

```
public class CompositeAdEventListener implements AdEvent.AdEventListener {
    private final List<AdEvent.AdEventListener> listeners;

    public CompositeAdEventListener(List<AdEvent.AdEventListener> listeners) {
        this.listeners = listeners;
    }

    @Override
    public void onAdEvent(AdEvent adEvent) {
        for (AdEvent.AdEventListener listener : listeners) {
            listener.onAdEvent(adEvent);
        }
    }
}
```

Hope this helps others. Also, having `addAdEventListener` with Exoplayer IMA extension would be a plus. Closing this issue now. Thanks.
